Add support for log emitter provider

elevenzhan opened the pull request:
In this CL:
* add default noop log emitter, guaranteed to be consistent with tracer and metrics
* add logEmitterProvider api
* add unit test cases

anuraaga reviewed (commented):
@jack-berg Can you check out this PR?

anuraaga commented:
@trask Do I understand correctly that your PR is the same as this API?

https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4917

trask commented:
> @trask Do I understand correctly that your PR is the same as this API?
> 
> [open-telemetry/opentelemetry-java-instrumentation#4917](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4917)

I don't think so. open-telemetry/opentelemetry-java-instrumentation#4917 introduces a way for instrumentation to avoid direct dependency on SDK, which we need for javaagent bridging.

jack-berg commented on sdk/logs/src/main/java/io/opentelemetry/sdk/logs/GlobalLogEmitterProvider.java:
```diff
@@ -0,0 +1,28 @@
+
+package io.opentelemetry.sdk.logs;
+
+/** This class is a temporary solution until log SDK is marked stable. */
```
Two thoughts:

1. Do we want a global accessor for the log sdk? There's not a precedent for providing global accessor for SDK components. `GlobalMeterProvider` provides access to `MeterProvider`, not `SdkMeterProvider`. `GlobalOpenTelemetry` provides access to `OpenTelemetry`, not `OpenTelemetrySdk`. Who would use this?

2.  Related to the javadoc: because there are no plans for a log api, this likely isn't a temporary solution.

elevenzhan replied:
Ok, I see and It has been deleted; this can be defined in my business code in a private way

jack-berg commented on sdk/logs/src/main/java/io/opentelemetry/sdk/logs/DefaultLogEmitter.java:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  // Noop implementation of LogEmitter.Builder.
+  private static final class NoopSpanBuilder implements LogBuilder {
```
`NoopSpanBuilder` => `NoopLogBuilder`

elevenzhan replied:
done

jack-berg commented on sdk/logs/src/main/java/io/opentelemetry/sdk/logs/DefaultLogEmitter.java:
```diff
@@ -0,0 +1,80 @@
+
+  @Override
+  public LogBuilder logBuilder() {
+    return new NoopSpanBuilder();
```
Can use a singleton here instead of instantiating a new instance each time.

elevenzhan replied:
done

anuraaga commented:
Hi @elevenzhan - we are currently working on a logs appender API in the instrumentation API here

https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4917

This is because we only expect it to be used by log appenders, not by end users. Will that satisfy your use case?

elevenzhan commented:
> Hi @elevenzhan - we are currently working on a logs appender API in the instrumentation API here
> 
> [open-telemetry/opentelemetry-java-instrumentation#4917](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4917)
> 
> This is because we only expect it to be used by log appenders, not by end users. Will that satisfy your use case?

Hi @anuraaga - I simplified the implementation and removed global provider; javaagent is a greatest way to resolve logs export, but in some unusable scenarios, It cannot overwrite these scenes; 
this CR provides a LogEmitterProvider api and noop implementation, that can be used anywhere who want to use log exporter, not just in javaagent

anuraaga commented:
> that can be used anywhere who want to use log exporter, not just in javaagent

The logs appender API in the java-instrumentation repo is just a library, it will be used for instrumentation both within the javaagent and without. So I think this one is mostly a dupe of that API

elevenzhan commented:
> > that can be used anywhere who want to use log exporter, not just in javaagent
> 
> The logs appender API in the java-instrumentation repo is just a library, it will be used for instrumentation both within the javaagent and without. So I think this one is mostly a dupe of that API

thanks @anuraaga, I see instrumentation-api-appender is consistent with current MR, so I will close this merge
